File: code/backend/python/app/core/services/match/time_start_recovery_service.py
"""Step 26 — time-start recovery, class bonuses & location counters.

Mirrors the Java ``TimeStartRecoveryService``. On every time-start the engine
recovers each character's stats based on their location safety, applies the
class bonuses, clamps to the caps, and decrements location time counters
(seeding missing rows for occupied counter-locations).

Step 33: when a counter reaches zero the location's ``id_event_if_counter_zero`` is
handed **up** rather than merely logged, together with the
``id_event_if_character_start_time`` of every occupied location. This service collects
those events; it does not run them. The event engine sits *above* it in the wiring (an
event can force a time end, which comes back here), so executing them from inside would
close a dependency cycle — ``TimeAdvancementService`` runs the list once this returns.

Recovery rules (P = location.secure_param + difficulty.energy; safe when
secure_param > 0):
  * safe   -> energy += DEX + P, life += COS + secure_param, sadness -= INT + secure_param
  * unsafe -> energy += difficulty.energy only (no DEX, no secure_param)
"""
from typing import Any, Dict, List

from app.core.models.match import location_entry_models as lem
from app.core.models.match.location_entry_models import PendingAutomaticEvent
from app.core.models.match.time_models import RecoveryItem, TimeStartOutcome
from app.core.ports.match.edge_state_ports import EdgeStateStorePort
from app.core.ports.match.time_ports import TimeStorePort
from app.core.services.match import edge_state_evaluator


class TimeStartRecoveryService:
    def __init__(self, store: TimeStorePort, edge_store: EdgeStateStorePort = None) -> None:
        self.store = store
        self.edge_store = edge_store

    def apply_at_time_start(self, id_match: int) -> TimeStartOutcome:
        ctx = self.store.load_recovery_context(id_match)
        if ctx is None:
            return TimeStartOutcome([], [])
        characters = self.store.find_recovery_characters(id_match) or []
        if not characters:
            return TimeStartOutcome([], [])

        id_story = ctx["id_story"]
        difficulty_energy = _nz(ctx.get("difficulty_energy"))
        current_clock = _nz(ctx.get("current_clock"))

        safety_by_location: Dict[int, Dict[str, Any]] = {
            s["id_location"]: s for s in self.store.find_location_safety(id_story)
        }
        all_bonuses = self.store.find_class_bonuses(id_story) or []
        state_rows = self.store.find_state_locations(id_match)
        counter_by_location: Dict[int, int] = {
            s["id_location"]: _nz(s.get("clock_counter")) for s in state_rows
        }
        flag_by_location: Dict[int, int] = {
            s["id_location"]: _nz(s.get("flag_already_actived")) for s in state_rows
        }

        # Build the set of locations occupied by characters.
        occupied = {c["id_location"] for c in characters if c.get("id_location") is not None}

        # 1a. Re-seed occupied locations whose existing counter is 0, not yet activated,
        #     but the story definition now carries counter_time > 0. Fixes matches
        #     created before counter_time was added to the location.
        for id_location in occupied:
            existing = counter_by_location.get(id_location)
            if existing is None or existing > 0:
                continue
            if flag_by_location.get(id_location, 0) != 0:
                continue
            s = safety_by_location.get(id_location)
            counter_time = _nz(s.get("counter_time")) if s else 0
            if counter_time <= 0:
                continue
            self.store.update_state_location_counter(id_match, id_location, counter_time)
            counter_by_location[id_location] = counter_time

        # 1b. Seed missing state-location rows for occupied counter-locations.
        for id_location in occupied:
            if id_location in counter_by_location:
                continue
            s = safety_by_location.get(id_location)
            counter_time = _nz(s.get("counter_time")) if s else 0
            if counter_time > 0:
                self.store.insert_state_location(id_match, id_location, counter_time)
                counter_by_location[id_location] = counter_time

        # 2. Recover each character.
        recaps: List[RecoveryItem] = []
        coma_after: List[bool] = []
        for c in characters:
            s = safety_by_location.get(c.get("id_location"))
            secure_param = _nz(s.get("secure_param")) if s else 0
            safe = secure_param > 0
            p = secure_param + difficulty_energy
            bonuses = [b for b in all_bonuses if b.get("id_class") == c.get("id_class")]
            energy, life, sad, sad_unclamped = compute_recovery(
                _nz(c.get("dexterity")), _nz(c.get("intelligence")), _nz(c.get("constitution")),
                _nz(c.get("energy")), _nz(c.get("life")), _nz(c.get("sad")),
                _nz(c.get("energy_max")), _nz(c.get("life_max")), _nz(c.get("sad_max")),
                safe, p, difficulty_energy,
                _sum_bonus(bonuses, "energy"), _sum_bonus(bonuses, "life"), _sum_bonus(bonuses, "sad"),
            )
            # A recovery can still push a character over an edge: a positive class sad bonus
            # raises sadness, and an unsafe location never heals. Evaluate before the write so
            # the corrected values land in the same UPDATE.
            v = edge_state_evaluator.evaluate(edge_state_evaluator.CharacterState(
                c["id"], life, sad_unclamped, _nz(c.get("sad_max")),
                _nz(c.get("constitution")), bool(c.get("is_coma"))))

            energy_delta = energy - _nz(c.get("energy"))
            life_delta = v.life_after - _nz(c.get("life"))
            sad_delta = v.sad_after - _nz(c.get("sad"))
            self.store.update_character_stats(id_match, c["id"], energy, v.life_after,
                                              v.sad_after)
            self.store.log_recovery(
                id_match, c["id"],
                f"recovery safe={safe} p={p} dEnergy={energy_delta} dLife={life_delta} dSad={sad_delta}",
            )
            edge_state_evaluator.persist(self.edge_store, id_match, v, current_clock, None)
            recaps.append(RecoveryItem(c["uuid"], energy_delta, life_delta, sad_delta))

            # v0.30.1 — a comatose character who rested in a safe location wakes. Safe recovery
            # has already lifted life above zero (life += COS + secure_param, both >= 1), so the
            # guard cannot leave it awake-but-dead to re-coma next clock. Independent of the
            # others: one character waking does not require the rest of the location to wake.
            still_coma = v.coma_triggered or bool(c.get("is_coma"))
            if bool(c.get("is_coma")) and safe and v.life_after > 0:
                self.edge_store.clear_coma(id_match, c["id"])
                self.edge_store.log_edge_state(
                    id_match, c["id"], None, current_clock,
                    f"{edge_state_evaluator.MSG_COMA_RECOVERED} {c['id']}")
                still_coma = False
            coma_after.append(still_coma)

        if edge_state_evaluator.all_in_coma(coma_after):
            edge_state_evaluator.log_all_player_coma(self.edge_store, id_match, None,
                                                     current_clock)

        # 3. Decrement counters; flag zeros and collect the events they owe.
        #    A counter is a ONE-SHOT FUSE: `current <= 0` skips an exhausted one and
        #    mark_state_location_activated latches it, so the event fires exactly once per
        #    match and the counter never restarts.
        pending_events: List[PendingAutomaticEvent] = []
        for id_location, current in counter_by_location.items():
            current = _nz(current)
            if current <= 0:
                continue
            nxt = current - 1
            self.store.update_state_location_counter(id_match, id_location, nxt)
            if nxt == 0:
                s = safety_by_location.get(id_location)
                pending = s.get("id_event_if_counter_zero") if s else None
                msg = f"counter reached zero at location {id_location}"
                if pending is not None:
                    msg += f"; pending event {pending}"
                self.store.log_counter_zero(id_match, id_location, pending,
                                            current_clock, msg)
                self.store.mark_state_location_activated(id_match, id_location)
                _add_pending(pending_events, lem.TRIGGER_COUNTER_ZERO, id_location,
                             pending, _nominal_actor(characters, id_location), s)

        # 4. Step 33 — a time unit BEGINNING where a character stands is its own trigger,
        #    independent of any counter. One entry per occupied location, not per
        #    character: the event describes the place, and the nominal actor is who it
        #    happens to.
        for id_location in occupied:
            s = safety_by_location.get(id_location)
            if not s:
                continue
            _add_pending(pending_events, lem.TRIGGER_CHARACTER_START_TIME, id_location,
                         s.get("id_event_if_character_start_time"),
                         _nominal_actor(characters, id_location), s)

        # Deterministic across locations: priority_automatic_event first, then location id.
        pending_events.sort(key=lambda p: (p.priority, p.id_location))

        return TimeStartOutcome(recaps, pending_events)


def _add_pending(out: List[PendingAutomaticEvent], trigger: str, id_location: int,
                 id_event, id_actor_character, safety) -> None:
    """Skips a null or non-positive event id — an unauthored trigger is not a trigger."""
    if not id_event or id_event <= 0:
        return
    priority = _nz(safety.get("priority_automatic_event")) if safety else 0
    out.append(PendingAutomaticEvent(trigger, id_location, int(id_event),
                                     id_actor_character, priority))


def _nominal_actor(characters: List[Dict[str, Any]], id_location: int):
    """The lowest-id character standing in a location, or None when nobody is.

    An automatic location event belongs to the place, not to a player, but its effects
    still have to resolve ``target = ONLY_ONE`` against somebody and ``target = ALL``
    against everyone *there*. Picking the lowest id makes that choice deterministic;
    picking nobody, when the place is empty, is equally correct — the world still changes,
    it just changes around no one.
    """
    ids = [c["id"] for c in characters
           if c.get("id_location") == id_location and c.get("id") is not None]
    return min(ids) if ids else None


def compute_recovery(dexterity: int, intelligence: int, constitution: int,
                     energy: int, life: int, sad: int,
                     energy_max: int, life_max: int, sad_max: int,
                     safe: bool, p: int, difficulty_energy: int,
                     bonus_energy: int, bonus_life: int, bonus_sad: int):
    """Pure recovery math: safe/unsafe formula + class bonuses + clamping."""
    secure_param = p - difficulty_energy
    new_energy = energy + dexterity + p if safe else energy + difficulty_energy
    new_life = life
    new_sad = sad
    if safe:
        new_life = life + constitution + secure_param
        new_sad = sad - (intelligence + secure_param)
    new_energy += bonus_energy
    new_life += bonus_life
    new_sad += bonus_sad
    return (
        _clamp(new_energy, 0, energy_max),
        _clamp(new_life, 0, life_max),
        _clamp(new_sad, 0, sad_max),
        new_sad,
    )


def _clamp(value: int, low: int, high: int) -> int:
    if high < low:
        return low
    return max(low, min(high, value))


def _sum_bonus(bonuses: List[Dict[str, Any]], stat: str) -> int:
    return sum(_nz(b.get("value")) for b in bonuses
               if str(b.get("statistic", "")).lower() == stat)


def _nz(value) -> int:
    return int(value) if value is not None else 0

File: code/backend/java/core/src/main/java/games/paths/core/service/match/TimeStartRecoveryService.java
package games.paths.core.service.match;

import games.paths.core.port.match.EdgeStateStorePort;
import games.paths.core.port.match.LocationEntryPort;
import games.paths.core.port.match.LocationEntryPort.PendingAutomaticEvent;
import games.paths.core.port.match.RecoveryStorePort;
import games.paths.core.port.match.RecoveryStorePort.ClassBonusView;
import games.paths.core.port.match.RecoveryStorePort.LocationSafety;
import games.paths.core.port.match.RecoveryStorePort.RecoveryCharacter;
import games.paths.core.port.match.RecoveryStorePort.RecoveryMatchContext;
import games.paths.core.port.match.RecoveryStorePort.StateLocationView;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * TimeStartRecoveryService - applies per-character stat recovery, class bonuses
 * and location time-counter decrements at every time-start (Step 26).
 *
 * <p>Recovery rules (P = location.secure_param + difficulty.energy; a location is
 * <em>safe</em> when secure_param &gt; 0):</p>
 * <ul>
 *   <li><b>safe</b> → energy += DEX + P, life += COS + secure_param, sadness -= INT + secure_param</li>
 *   <li><b>unsafe</b> → energy += difficulty.energy only (no DEX, no secure_param)</li>
 * </ul>
 * Class bonuses (list_classes_bonus for the character's class) are added on top,
 * then every value is clamped: 0 ≤ energy ≤ energy_max, 0 ≤ life ≤ life_max,
 * 0 ≤ sad ≤ sad_max.
 *
 * <p>Location counters: a location occupied by a player and carrying
 * counter_time &gt; 0 but lacking a gaming_state_locations row is seeded; every
 * existing counter &gt; 0 is decremented; when it reaches zero the location's
 * id_event_if_counter_zero is logged as pending (execution wired in Step 29).</p>
 *
 * <p>Step 30: every recovered character then goes through {@link EdgeStateEvaluator}. A
 * recovery is not only healing — an unsafe location restores no life at all, and a positive
 * class {@code sad} bonus can push sadness over its cap — so the same overflow and coma rules
 * that guard event execution apply here too.</p>
 *
 * <p>See {@code documentation_v0/Step26_TimeStartRecovery.md} and
 * {@code documentation_v0/Step30_EdgeStates.md}.</p>
 */
public class TimeStartRecoveryService {

    private final RecoveryStorePort store;
    private final EdgeStateStorePort edgeStore;

    public TimeStartRecoveryService(RecoveryStorePort store, EdgeStateStorePort edgeStore) {
        this.store = store;
        this.edgeStore = edgeStore;
    }

    /**
     * Run the full time-start recovery sequence for a match: the per-character recap
     * (deltas applied) and, since Step 33, the automatic events this time-start owes —
     * the counters that reached zero and the {@code id_event_if_character_start_time}
     * of every occupied location.
     *
     * <p>It collects those events rather than running them. The event engine sits
     * <em>above</em> this service in the wiring (an event can force a time end, which
     * comes back here), so executing them from inside would close a dependency cycle.
     * {@code TimeAdvancementService} runs the list once this returns.</p>
     *
     * <p>Empty when the match has no resolvable story context or no characters.</p>
     */
    public TimeStartOutcome applyAtTimeStart(long idMatch) {
        Optional<RecoveryMatchContext> ctxOpt = store.loadContext(idMatch);
        if (ctxOpt.isEmpty()) {
            return TimeStartOutcome.none();
        }
        RecoveryMatchContext ctx = ctxOpt.get();
        List<RecoveryCharacter> characters = store.findCharacters(idMatch);
        if (characters.isEmpty()) {
            return TimeStartOutcome.none();
        }

        Map<Long, LocationSafety> safetyByLocation = new HashMap<>();
        for (LocationSafety s : store.findLocationSafety(ctx.idStory())) {
            safetyByLocation.put(s.idLocation(), s);
        }
        List<ClassBonusView> allBonuses = store.findClassBonuses(ctx.idStory());

        Map<Long, Integer> counterByLocation = new HashMap<>();
        Map<Long, Integer> flagByLocation = new HashMap<>();
        for (StateLocationView sl : store.findStateLocations(idMatch)) {
            counterByLocation.put(sl.idLocation(), sl.clockCounter());
            flagByLocation.put(sl.idLocation(), sl.flagAlreadyActived());
        }

        // Build the set of locations currently occupied by characters.
        Set<Long> occupied = new HashSet<>();
        for (RecoveryCharacter c : characters) {
            if (c.idLocation() != null) occupied.add(c.idLocation());
        }

        // 1a. Re-seed occupied locations where the existing row has clockCounter = 0
        //     and has never been activated, but the story definition now carries
        //     counterTime > 0. This fixes matches created before counter_time was set.
        for (Long idLocation : occupied) {
            Integer existing = counterByLocation.get(idLocation);
            if (existing == null || existing > 0) continue;
            if (flagByLocation.getOrDefault(idLocation, 0) != 0) continue;
            LocationSafety s = safetyByLocation.get(idLocation);
            int counterTime = s == null ? 0 : nz(s.counterTime());
            if (counterTime <= 0) continue;
            store.updateStateLocationCounter(idMatch, idLocation, counterTime);
            counterByLocation.put(idLocation, counterTime);
        }

        // 1b. Seed missing state-location rows for occupied locations with a counter.
        for (Long idLocation : occupied) {
            if (counterByLocation.containsKey(idLocation)) continue;
            LocationSafety s = safetyByLocation.get(idLocation);
            int counterTime = s == null ? 0 : nz(s.counterTime());
            if (counterTime > 0) {
                store.insertStateLocation(idMatch, idLocation, counterTime);
                counterByLocation.put(idLocation, counterTime);
            }
        }

        // 2. Recover each character (recovery + class bonus + clamp + Step 30 edge states).
        List<RecoveryRecap> recaps = new ArrayList<>();
        List<Boolean> comaAfter = new ArrayList<>();
        for (RecoveryCharacter c : characters) {
            LocationSafety s = c.idLocation() == null ? null : safetyByLocation.get(c.idLocation());
            int secureParam = s == null ? 0 : nz(s.secureParam());
            boolean safe = secureParam > 0;
            int p = secureParam + ctx.difficultyEnergy();

            List<ClassBonusView> bonuses = bonusesForClass(allBonuses, c.idClass());
            StatTriple result = computeRecovery(
                    c.dexterity(), c.intelligence(), c.constitution(),
                    c.energy(), c.life(), c.sad(),
                    c.energyMax(), c.lifeMax(), c.sadMax(),
                    safe, p, ctx.difficultyEnergy(),
                    sumBonus(bonuses, "energy"),
                    sumBonus(bonuses, "life"),
                    sumBonus(bonuses, "sad"));

            // A recovery can still push a character over an edge: a positive class sad bonus
            // raises sadness, and an unsafe location never heals. Evaluate before the write so
            // the corrected values land in the same UPDATE.
            EdgeStateEvaluator.Verdict verdict = EdgeStateEvaluator.evaluate(
                    new EdgeStateEvaluator.CharacterState(
                            c.id(), result.life(), result.sadUnclamped(), c.sadMax(),
                            c.constitution(), c.isComa()));

            int energyDelta = result.energy() - c.energy();
            int lifeDelta = verdict.lifeAfter() - c.life();
            int sadDelta = verdict.sadAfter() - c.sad();
            store.updateCharacterStats(idMatch, c.id(),
                    result.energy(), verdict.lifeAfter(), verdict.sadAfter());
            store.logRecovery(idMatch, c.id(),
                    "recovery safe=" + safe + " p=" + p
                            + " dEnergy=" + energyDelta + " dLife=" + lifeDelta + " dSad=" + sadDelta);
            EdgeStateEvaluator.persist(edgeStore, idMatch, verdict, ctx.currentClock(), null);
            recaps.add(new RecoveryRecap(c.uuid(), energyDelta, lifeDelta, sadDelta));

            // v0.30.1 — a comatose character who rested in a safe location wakes. Safe recovery
            // has already lifted its life above zero (life += COS + secure_param, both ≥ 1), so
            // the guard cannot leave it awake-but-dead to re-coma next clock. Independent of the
            // others: one character waking does not require the rest of the location to wake.
            boolean stillComa = verdict.comaTriggered() || c.isComa();
            if (c.isComa() && safe && verdict.lifeAfter() > 0) {
                edgeStore.clearComa(idMatch, c.id());
                edgeStore.logEdgeState(idMatch, c.id(), null, ctx.currentClock(),
                        EdgeStateStorePort.MSG_COMA_RECOVERED + " " + c.id());
                stillComa = false;
            }
            comaAfter.add(stillComa);
        }

        // The whole party can go under during a recovery just as during an event. The row is
        // written here; running the story epilogue is the event engine's job, which owns the
        // chain runner and a result object to carry a card back in.
        if (EdgeStateEvaluator.allInComa(comaAfter)) {
            edgeStore.logEdgeState(idMatch, null, null, ctx.currentClock(),
                    EdgeStateStorePort.MSG_ALL_PLAYER_COMA + " " + idMatch);
        }

        // 3. Decrement location counters; flag zeros and collect the events they owe.
        //    A counter is a ONE-SHOT FUSE: `current <= 0` skips an exhausted one and
        //    markStateLocationActivated latches it, so the event fires exactly once per
        //    match and the counter never restarts.
        List<PendingAutomaticEvent> pending = new ArrayList<>();
        for (Map.Entry<Long, Integer> e : counterByLocation.entrySet()) {
            long idLocation = e.getKey();
            int current = nz(e.getValue());
            if (current <= 0) continue;
            int next = current - 1;
            store.updateStateLocationCounter(idMatch, idLocation, next);
            if (next == 0) {
                LocationSafety s = safetyByLocation.get(idLocation);
                Integer pendingEvent = s == null ? null : s.idEventIfCounterZero();
                store.logCounterZero(idMatch, idLocation, pendingEvent, ctx.currentClock(),
                        "counter reached zero at location " + idLocation
                                + (pendingEvent != null ? "; pending event " + pendingEvent : ""));
                store.markStateLocationActivated(idMatch, idLocation);
                addPending(pending, LocationEntryPort.TRIGGER_COUNTER_ZERO, idLocation,
                        pendingEvent, nominalActor(characters, idLocation), s);
            }
        }

        // 4. Step 33 — a time unit BEGINNING where a character stands is its own trigger,
        //    independent of any counter. One entry per occupied location, not per character:
        //    the event describes the place, and the nominal actor is who it happens to.
        for (Long idLocation : occupied) {
            LocationSafety s = safetyByLocation.get(idLocation);
            if (s == null) continue;
            addPending(pending, LocationEntryPort.TRIGGER_CHARACTER_START_TIME, idLocation,
                    s.idEventIfCharacterStartTime(), nominalActor(characters, idLocation), s);
        }

        // Deterministic across locations: priority_automatic_event first, then location id.
        pending.sort(Comparator.comparingInt(PendingAutomaticEvent::priority)
                .thenComparingLong(PendingAutomaticEvent::idLocation));

        return new TimeStartOutcome(recaps, pending);
    }

    /** Skips a null or non-positive event id — an unauthored trigger is not a trigger. */
    private static void addPending(List<PendingAutomaticEvent> out, String trigger, long idLocation,
                                   Integer idEvent, Long idActorCharacter, LocationSafety safety) {
        if (idEvent == null || idEvent <= 0) {
            return;
        }
        int priority = safety == null || safety.priorityAutomaticEvent() == null
                ? 0
                : safety.priorityAutomaticEvent();
        out.add(new PendingAutomaticEvent(trigger, idLocation, idEvent.longValue(),
                idActorCharacter, priority));
    }

    /**
     * The lowest-id character standing in a location, or null when nobody is.
     *
     * <p>An automatic location event belongs to the place, not to a player, but its effects
     * still have to resolve {@code target = ONLY_ONE} against somebody and {@code target =
     * ALL} against everyone <em>there</em>. Picking the lowest id makes that choice
     * deterministic; picking nobody, when the place is empty, is equally correct — the world
     * still changes, it just changes around no one.</p>
     */
    private static Long nominalActor(List<RecoveryCharacter> characters, long idLocation) {
        Long lowest = null;
        for (RecoveryCharacter c : characters) {
            if (c.idLocation() == null || c.idLocation() != idLocation) {
                continue;
            }
            if (lowest == null || c.id() < lowest) {
                lowest = c.id();
            }
        }
        return lowest;
    }

    // ── pure recovery math (unit-tested directly) ────────────────────────────

    /**
     * Compute the recovered (energy, life, sad) for a character, applying the
     * safe/unsafe recovery formula, the class bonuses and the stat caps.
     */
    public static StatTriple computeRecovery(int dexterity, int intelligence, int constitution,
                                             int energy, int life, int sad,
                                             int energyMax, int lifeMax, int sadMax,
                                             boolean safe, int p, int difficultyEnergy,
                                             int bonusEnergy, int bonusLife, int bonusSad) {
        int secureParam = p - difficultyEnergy;
        int newEnergy = safe ? energy + dexterity + p : energy + difficultyEnergy;
        int newLife = life;
        int newSad = sad;
        if (safe) {
            newLife = life + constitution + secureParam;
            newSad = sad - (intelligence + secureParam);
        }
        newEnergy += bonusEnergy;
        newLife += bonusLife;
        newSad += bonusSad;
        return new StatTriple(
                clamp(newEnergy, 0, energyMax),
                clamp(newLife, 0, lifeMax),
                clamp(newSad, 0, sadMax),
                newSad);
    }

    public static int clamp(int value, int min, int max) {
        if (max < min) {
            return min;
        }
        return Math.max(min, Math.min(max, value));
    }

    private static List<ClassBonusView> bonusesForClass(List<ClassBonusView> all, Long idClass) {
        if (idClass == null) {
            return List.of();
        }
        List<ClassBonusView> out = new ArrayList<>();
        for (ClassBonusView b : all) {
            if (b.idClass() == idClass) {
                out.add(b);
            }
        }
        return out;
    }

    private static int sumBonus(List<ClassBonusView> bonuses, String stat) {
        int total = 0;
        for (ClassBonusView b : bonuses) {
            if (stat.equalsIgnoreCase(b.statistic())) {
                total += b.value();
            }
        }
        return total;
    }

    private static int nz(Integer v) {
        return v == null ? 0 : v;
    }

    /**
     * Final recovered stats. {@code sadUnclamped} is the raw sadness before the cap, which the
     * Step 30 overflow rule reads: a positive class {@code sad} bonus can push a character over
     * the cap during what is nominally a recovery.
     */
    public record StatTriple(int energy, int life, int sad, int sadUnclamped) {
    }

    /** Per-character recovery summary surfaced in the time-advance response. */
    public record RecoveryRecap(String characterUuid, int energyDelta, int lifeDelta, int sadDelta) {
    }

    /**
     * What one time-start produced: the stat deltas, and the automatic events it owes,
     * already ordered by {@code priority_automatic_event} then location id.
     */
    public record TimeStartOutcome(List<RecoveryRecap> recovery,
                                   List<PendingAutomaticEvent> pending) {

        public static TimeStartOutcome none() {
            return new TimeStartOutcome(List.of(), List.of());
        }
    }
}
